src/components/shared/input.tsx

```diff
@@ -1,51 +1,97 @@
+import { cn } from '../../lib/utils.ts';
+import CircleXMarkIcon from '../../assets/icons/circle-x-mark-icon.tsx';
+import EyeSlashIcon from '../../assets/icons/eye-slash-icon.tsx';
+import EyeIcon from '../../assets/icons/eye-icon.tsx';
 
-// import { cva, VariantProps } from 'class-variance-authority';
-// const variants = cva(
-//   'flex items-center rounded-lg border-[1px] p-2 text-sm font-normal text-text-001 placeholder-text-003 border-brand-gray-light',
-//   {
-//     variants: {
-//       state: {
-//         empty: 'border-brand-gray-heavy',
-//         focused: 'border-brand-blue-heavy',
-//         filled: 'bg-brand-gray-light',
-//       },
-//     },
-//     defaultVariants: {
-//       state: 'empty',
-//     },
-//   },
-// );
+import { cva, VariantProps } from 'class-variance-authority';
+
+const variants = cva(
+  'w-full rounded-md p-2.5 text-sm font-medium outline-2 placeholder:font-normal placeholder:text-crews-g04 focus:bg-crews-b02 focus:outline focus:outline-crews-b04',
+  {
+    variants: {
+      state: {
+        empty: 'bg-crews-g01',
+        filled: 'outline outline-crews-g01 bg-crews-w01',
+        error:
+          'outline outline-crews-r02 bg-crews-r01 focus:outline-crews-r02 focus:bg-crews-r01 animation-shake',
+      },
+    },
+    defaultVariants: {
+      state: 'empty',
+    },
+  },
+);
 
 type InputProps = Omit<
   React.InputHTMLAttributes<HTMLInputElement>,
   'className'
 > & {
   className?: string;
   registerReturns: UseFormRegisterReturn;
-  handleXClick: React.MouseEventHandler<HTMLButtonElement>;
-};
-// & VariantProps<typeof variants>;
+  clearInput: () => void;
+} & VariantProps<typeof variants>;
 
 const Input = ({
   className,
   registerReturns,
-  handleXClick,
+  clearInput,
+  state,
   ...props
 }: InputProps) => {
+  const [focus, setFocus] = useState<boolean>(false);
+
+  const [customInputType, setCustomInputType] = useState<string>(
+    props.type || 'text',
+  );
+
+  const toggleCustomInputType: React.MouseEventHandler<HTMLButtonElement> = (
+    e,
+  ) => {
+    e.preventDefault();
+    setCustomInputType((prev) => (prev === 'text' ? 'password' : 'text'));
+  };
+
+  const handleXClick: React.MouseEventHandler<HTMLButtonElement> = (e) => {
+    e.preventDefault();
+    clearInput();
+  };
+
   return (
     <div
-      className={className}
-      // className={cn(
-      //   variants({ state }),
-      //   className,
-      // )}
+      className={cn(
+        'relative',
+        className,
+        state === 'error' ? 'animate-shake' : '',
+      )}
     >
-      <input {...registerReturns} {...props} />
-      <button type="button" onClick={handleXClick}>
-        x
-      </button>
+      <input
+        className={cn(variants({ state }))}
+        {...registerReturns}
+        {...props}
+        type={customInputType}
+        onFocus={() => setFocus(true)}
+        onBlur={() => setFocus(false)}
+        autoComplete="off"
+      />
```

Comment:
`props`를 스프레드로 넘겨줄 시에 상단에 위치하게 되면 `onFocus`와 같은 이벤트를 props로 넘겼을 시에, 가장 마지막에 선언된 `onFocus`가 실행되기 때문에 가장 아래에 위치하는 것이 좋다고 생각이 듭니다!

```suggestion
      <input
        className={cn(variants({ state }))}
        {...registerReturns}
        type={customInputType}
        onFocus={() => setFocus(true)}
        onBlur={() => setFocus(false)}
        autoComplete="off"
         {...props}
      />
```

Reply:
앗 저는 오히려 그래서 `onFocus`가 prop에 의해 덮어씌워지면 안된다고 생각해서 하단 부에 위치시켰습니다..!

차라리

```ts
Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  'className'
> 
```
에 `onFocus`를 추가하는 것이 의도치 않은 동작의 발생 가능성을 낮출 것 같은데 어떻게 생각하시나요? 🧐

Reply:
아앗 그렇군요! 저는 반대로 onFocus와 같은 부분들도 override 가능한 걸로 이해하고 코멘트달았는데 `Omit`에 추가하는 거 좋은 것 같습니다!